lib/msf/core/exploit/browser_autopwnv2.rb

```diff
@@ -0,0 +1,834 @@
+      xploit.datastore['Retries']     = datastore['Retries'] if datastore['Retries']
+      xploit.datastore['SSL']         = datastore['SSL'] if datastore['SSL']
+      xploit.datastore['SSLVersion']  = datastore['SSLVersion'] if datastore['SSLVersion']
+      xploit.datastore['LHOST']       = get_payload_lhost
+
+      # Set options only configurable by BAP.
+      xploit.datastore['DisablePayloadHandler'] = true
+      xploit.datastore['NoteTypePrefix']        = note_type_prefix
+      xploit.datastore['URIPATH']               = "/#{assign_module_resource}"
+      xploit.datastore['MODULEOWNER']           = Msf::Exploit::Remote::BrowserAutopwnv2
+    end
+
+
+    # Checks if a resource is already taken or not.
+    #
+    # @param resource [String] The resource to check.
+    # @return [TrueClass] Resource is taken.
+    # @return [FalseClass] Resource is not taken.
+    def is_resource_taken?(resource)
+      taken = false
+
+      bap_exploits.each do |m|
+        return true if m.datastore['URIPATH'] == resource
+      end
+
+      taken
+    end
+
+
+    # Returns a unique resource path.
+    #
+    # @return [String] A unique resource path.
+    def assign_module_resource
+      resource = ''
+      while
+        resource = Rex::Text.rand_text_alpha(rand(10) + 4)
+        break unless is_resource_taken?(resource)
+      end
+
+      resource
+    end
+
+
+    # Modifies @bap_exploits by sorting. The newest and with the highest ranking goes on top.
+    # This method is part of what makes BAP smarter. However, the list rearranged by this exploit
+    # will not actually be the same exploit list served to every client. When a client a request,
+    # #get_suitable_exploits will generate another list that will actually be used by the client
+    # by going through what we have here, and filter out all the exploit modules that don't match
+    # the target's requirements.
+    #
+    # @see #bap_exploits The read-only attribute.
+    # @see #sort_date_in_group The method for sorting by disclosure date
+    # @see #sort_group_by_rank The method for sorting by rank
+    # @see #sort_bap_modules The method for breaking the module list into groups
+    # @see #finalize_sorted_modules The method for finalizing bap_exploits
+    # @see #get_suitable_exploits
+    # @return [void]
+    def sort_bap_exploits
+      bap_groups = group_bap_modules
+      bap_groups = sort_date_in_group(bap_groups)
+      bap_groups = sort_group_by_rank(bap_groups)
+      finalize_sorted_modules(bap_groups)
+    end
+
+
+    # Sorts a grouped module list by disclosure date.
+    #
+    # @param bap_groups [Hash] A grouped module list.
+    # @return [Hash] A hash with each module list sorted by disclosure date.
+    def sort_date_in_group(bap_groups)
+      bap_groups.each_pair do |ranking, module_list|
+        bap_groups[ranking] = module_list.sort_by {|m| Date.parse(m.disclosure_date.to_s)}.reverse
```

Comment:
Custom local modules would never get tested by Travis, so it may not hurt to be defensive here.